unconvert_test.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2019 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+func TestBinary(t *testing.T) {
+	exepath, cleanup := build(t)
+	defer cleanup()
+
+	output, err := exec.Command(exepath, "./testdata").CombinedOutput()
+	if err == nil {
+		t.Fatal("expected to quit with an error code")
+	}
+
+	got, err := ParseOutput("testdata", string(output))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expected, err := ParseDir("testdata")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	SortAnnotations(got)
+	SortAnnotations(expected)
+
+	if len(got) != len(expected) {
+		t.Errorf("different number of results: got %v expected %v", len(got), len(expected))
+	}
+
+	n := len(got)
+	if len(expected) < n {
+		n = len(expected)
+	}
+	for i, a := range got[:n] {
+		b := expected[i]
+		if a != b {
+			t.Errorf("got %q expected %q", a, b)
+		}
+	}
+}
+
+type Annotation struct {
+	File    string
+	Line    int
+	Message string
+}
+
+func (ann Annotation) String() string {
+	return fmt.Sprintf("%s:%d:0: %s", ann.File, ann.Line, ann.Message)
+}
+
+func SortAnnotations(xs []Annotation) {
+	sort.Slice(xs, func(i, k int) bool {
+		if xs[i].File == xs[k].File {
+			return xs[i].Line < xs[k].Line
+		}
+		return xs[i].File < xs[k].File
+	})
+}
+
+func ParseOutput(folder, output string) ([]Annotation, error) {
+	var all []Annotation
+	for _, line := range strings.Split(output, "\n") {
+		line = strings.TrimSpace(line)
+		if line == "" {
+			continue
+		}
+
+		folderStart := strings.Index(line, folder)
+		if folderStart < 0 {
+			continue
+		}
+
+		line = line[folderStart+len(folder)+1:]
+		tokens := strings.SplitN(line, ":", 4)
+		if len(tokens) != 4 {
+			continue
+		}
+
+		line, err := strconv.Atoi(tokens[1])
+		if err != nil {
+			return nil, err
+		}
+
+		all = append(all, Annotation{
+			File:    tokens[0],
+			Line:    line,
+			Message: strings.TrimSpace(tokens[3]),
+		})
+	}
+	return all, nil
+}
+
+func ParseDir(dir string) ([]Annotation, error) {
+	var all []Annotation
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+	for _, file := range files {
```

Comment:
I think you should include `if !strings.HasSuffix(file, ".go") { continue }` so that `README.md` and `foo.go~` files don't get caught up here.

Alternatively, maybe use `go/build` or `x/tools/go/packages` to load the "./testdata" directory and get their `GoFiles`.

Reply:
`go/build` etc. seemed overkill for this scenario, but added the extension check.